### PR TITLE
TaskStatusDB: `check_out_task`

### LINKED ISSUE #3: TaskStatusDB: Method to update task status
* `update_task_status(self, taskid, new_status, old_status)`

A couple concerns/questions:

* Do we need to pass the DB connection in here? This seems like it could be part of a more complicated sequence that we'd like to commit all at once.

### PR DESCRIPTION
**Pull request by dwhswenson**
Resolves #3. Supersedes #15.

This includes all of #14, and should only be reviewed after that.

### CODE REVIEW

File: exorcist/tests/test_taskstatusdb.py [hunk 1]
```diff
@@ -169,8 +189,72 @@ def test_get_all_tasks(self, request, fixture):
...
+                ('bar', TaskStatus.BLOCKED.value, None, 0, 3)},
         }[fixture]
         db = request.getfixturevalue(fixture)
         tasks = set(db.get_all_tasks())
         assert tasks == expected
+
+    def test_add_task(self, fresh_db):
+        # task without prerequisites
+        fresh_db.add_task("foo", requirements=[], max_tries=3)
+        expected_foo_task = ("foo", TaskStatus.AVAILABLE.value, None, 0, 3)
```
Comment (richardjgowers) [inline exorcist/tests/test_taskstatusdb.py:201]: what does the `..., None, 0, 3)`  data here mean?

File: exorcist/tests/test_taskstatusdb.py [hunk 2]
```diff
@@ -169,8 +189,72 @@ def test_get_all_tasks(self, request, fixture):
...
+        assert len(deps) == len(expected_deps)
+
+    def test_task_row_update_statement(self, loaded_db):
+        # TODO: I'm going to do this in a future PR
+        ...
+
+    def test_check_out_task(self, loaded_db):
+        taskid = loaded_db.check_out_task()
+        assert taskid == "foo"
+
```
Comment (richardjgowers) [inline exorcist/tests/test_taskstatusdb.py:242]: could we also test that the database doesn't repeatedly hand out the same task on multiple calls?  So we put in two tasks with no deps, and check that we don't get given 'foo' twice.

then similarly if we had foo, bar, baz where baz depends on foo and bar, can we check that only foo and bar get handed out?

File: exorcist/taskdb.py [hunk 3]
```diff
@@ -198,39 +234,245 @@ def _create_empty_db(metadata, engine):
...
         """
-        ...
+        if result.rowcount > 1: # -no-cov-
+            raise RuntimeError("Database seems to have more than 1 row with"
+                               f"taskid '{taskid}'. This should not happen.")
+        elif result.rowcount == 0:
+            raise NoStatusChange(f"Task '{taskid}' could not change from "
+                                 f"{old_status} to {status}")
 
     def check_out_task(self):
```
Comment (richardjgowers) [inline exorcist/taskdb.py:447]: docstring

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: exorcist/tests/test_taskstatusdb.py (hunk 1 above)
Comment (dwhswenson) [reply exorcist/tests/test_taskstatusdb.py:201]: This is the row of the `tasks` table. It returns a namedtuple (technically, an `sqla.Row`, which subclasses or wraps a namedtuple.

The entries are `(tasked, status, last_modified, tries, max_tries)`.

File: exorcist/taskdb.py (hunk 3 above)
Comment (dwhswenson) [reply exorcist/taskdb.py:447]: I was hoping we could count on inheritance from the ABC instead of having multiple copies of the same string?

**Review by richardjgowers — APPROVED** [review]
cool thanks, I didn't realise abc docstrings worked in that way, handy

File: exorcist/taskdb.py [hunk 4]
```diff
@@ -2,10 +2,13 @@
...
 import networkx as nx
 from .models import TaskStatus
+from datetime import datetime
 
-# remaining imports are for typing
-from typing import Optional, Iterable
+# imports for typing
+from typing import Optional, Iterable, Union
 from os import PathLike
+from sqlalchemy.sql.roles import StatementRole as SQLStatement
```
Comment (richardjgowers) [inline exorcist/taskdb.py:10]: another reason for splitting out the abc, now the abc depends on sqlalchemy thanks to this import

**Pull request merged**